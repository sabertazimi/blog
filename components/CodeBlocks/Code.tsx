import { classNames } from '@components/utils';
import type { HTMLProps, ReactElement } from 'react';
import styles from './Code.module.css';

interface Props extends HTMLProps<HTMLPreElement> {}

const getAlias = (language: string) => {
  switch (language) {
    case 'html':
      return 'HTML';
    case 'css':
      return 'CSS';
    case 'json':
      return 'JSON';
    case 'md':
    case 'markdown':
      return 'Markdown';
    case 'js':
    case 'javascript':
      return 'JavaScript';
    case 'ts':
    case 'typescript':
      return 'TypeScript';
    case 'jsx':
    case 'tsx':
      return 'React';
    default:
      return language?.charAt(0)?.toUpperCase() + language?.slice(1) || 'Code';
  }
};

const Code = ({ children, className, ...props }: Props): JSX.Element => {
  const childrenElement = children as ReactElement;
  const language = getAlias(
    (childrenElement?.props?.className as string)?.replace('language-', '')
  );

  return (
    <pre
      {...props}
      className={classNames(className, styles.code)}
      data-language={language}
    >
      {children}
    </pre>
  );
};

export default Code;
