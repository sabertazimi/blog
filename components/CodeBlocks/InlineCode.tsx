import cx from 'classnames'
import styles from './InlineCode.module.css'

interface Props {
  children?: string
  className?: string
}

function InlineCode({ children, className }: Props): JSX.Element {
  return (
    <code
      className={cx(
        className,
        styles.code,
        'dark:text-light dark:shadow-xl dark:shadow-primary',
      )}
    >
      {children}
    </code>
  )
}

export default InlineCode
