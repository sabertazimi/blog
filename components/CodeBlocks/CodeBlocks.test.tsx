import { render } from '@testing-library/react';
import Code from './Code';
import InlineCode from './InlineCode';

describe('InlineCode', () => {
  test('should render correctly (snapshot)', () => {
    const { container } = render(<InlineCode />);

    expect(container).toMatchSnapshot();
  });

  test('should render children correctly (snapshot)', () => {
    const { container } = render(<InlineCode>InlineCode</InlineCode>);

    expect(container).toMatchSnapshot();
  });

  test('should render raw code correctly (snapshot)', () => {
    const { container } = render(
      <InlineCode className="language-ts">InlineCode</InlineCode>
    );

    expect(container).toMatchSnapshot();
  });
});

describe('Code', () => {
  const languages = [
    'html',
    'css',
    'json',
    'md',
    'markdown',
    'js',
    'javascript',
    'ts',
    'typescript',
    'jsx',
    'tsx',
    'vue',
    'go',
    'java',
    'rust',
  ];

  test('should render correctly (snapshot)', () => {
    const { container } = render(<Code />);

    expect(container).toMatchSnapshot();
  });

  test('should render children correctly (snapshot)', () => {
    const { container } = render(
      <Code>
        <code>Code</code>
      </Code>
    );

    expect(container).toMatchSnapshot();
  });

  test.each(languages)(
    'should render language correctly (snapshot)',
    language => {
      const { container } = render(
        <Code>
          <code className={`language-${language}`}>Code</code>
        </Code>
      );

      expect(container).toMatchSnapshot();
    }
  );
});
