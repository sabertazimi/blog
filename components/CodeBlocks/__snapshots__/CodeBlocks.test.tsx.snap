// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Code should render children correctly (snapshot) 1`] = `
<div>
  <pre
    class="code"
    data-language="Code"
  >
    <code>
      Code
    </code>
  </pre>
</div>
`;

exports[`Code should render correctly (snapshot) 1`] = `
<div>
  <pre
    class="code"
    data-language="Code"
  />
</div>
`;

exports[`Code should render language correctly (snapshot) 1`] = `
<div>
  <pre
    class="code"
    data-language="HTML"
  >
    <code
      class="language-html"
    >
      Code
    </code>
  </pre>
</div>
`;

exports[`Code should render language correctly (snapshot) 2`] = `
<div>
  <pre
    class="code"
    data-language="CSS"
  >
    <code
      class="language-css"
    >
      Code
    </code>
  </pre>
</div>
`;

exports[`Code should render language correctly (snapshot) 3`] = `
<div>
  <pre
    class="code"
    data-language="JSON"
  >
    <code
      class="language-json"
    >
      Code
    </code>
  </pre>
</div>
`;

exports[`Code should render language correctly (snapshot) 4`] = `
<div>
  <pre
    class="code"
    data-language="Markdown"
  >
    <code
      class="language-md"
    >
      Code
    </code>
  </pre>
</div>
`;

exports[`Code should render language correctly (snapshot) 5`] = `
<div>
  <pre
    class="code"
    data-language="Markdown"
  >
    <code
      class="language-markdown"
    >
      Code
    </code>
  </pre>
</div>
`;

exports[`Code should render language correctly (snapshot) 6`] = `
<div>
  <pre
    class="code"
    data-language="JavaScript"
  >
    <code
      class="language-js"
    >
      Code
    </code>
  </pre>
</div>
`;

exports[`Code should render language correctly (snapshot) 7`] = `
<div>
  <pre
    class="code"
    data-language="JavaScript"
  >
    <code
      class="language-javascript"
    >
      Code
    </code>
  </pre>
</div>
`;

exports[`Code should render language correctly (snapshot) 8`] = `
<div>
  <pre
    class="code"
    data-language="TypeScript"
  >
    <code
      class="language-ts"
    >
      Code
    </code>
  </pre>
</div>
`;

exports[`Code should render language correctly (snapshot) 9`] = `
<div>
  <pre
    class="code"
    data-language="TypeScript"
  >
    <code
      class="language-typescript"
    >
      Code
    </code>
  </pre>
</div>
`;

exports[`Code should render language correctly (snapshot) 10`] = `
<div>
  <pre
    class="code"
    data-language="React"
  >
    <code
      class="language-jsx"
    >
      Code
    </code>
  </pre>
</div>
`;

exports[`Code should render language correctly (snapshot) 11`] = `
<div>
  <pre
    class="code"
    data-language="React"
  >
    <code
      class="language-tsx"
    >
      Code
    </code>
  </pre>
</div>
`;

exports[`Code should render language correctly (snapshot) 12`] = `
<div>
  <pre
    class="code"
    data-language="Vue"
  >
    <code
      class="language-vue"
    >
      Code
    </code>
  </pre>
</div>
`;

exports[`Code should render language correctly (snapshot) 13`] = `
<div>
  <pre
    class="code"
    data-language="Go"
  >
    <code
      class="language-go"
    >
      Code
    </code>
  </pre>
</div>
`;

exports[`Code should render language correctly (snapshot) 14`] = `
<div>
  <pre
    class="code"
    data-language="Java"
  >
    <code
      class="language-java"
    >
      Code
    </code>
  </pre>
</div>
`;

exports[`Code should render language correctly (snapshot) 15`] = `
<div>
  <pre
    class="code"
    data-language="Rust"
  >
    <code
      class="language-rust"
    >
      Code
    </code>
  </pre>
</div>
`;

exports[`InlineCode should render children correctly (snapshot) 1`] = `
<div>
  <code
    class="code"
  >
    InlineCode
  </code>
</div>
`;

exports[`InlineCode should render correctly (snapshot) 1`] = `
<div>
  <code
    class="code"
  />
</div>
`;

exports[`InlineCode should render raw code correctly (snapshot) 1`] = `
<div>
  <code
    class="language-ts"
  >
    InlineCode
  </code>
</div>
`;
